lib/chef/resource/locale.rb

```diff
@@ -36,56 +36,41 @@ class Locale < Chef::Resource
-            # Ubuntu 16.04 still uses /etc/default/locale
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang) } if ::File.exist?("/etc/default/locale")
-          end
+        if ::File.exist?("/etc/default/locale")
+          locale_file = "/etc/default/locale"
         elsif ::File.exist?("/etc/sysconfig/i18n")
-          locale_file_path = "/etc/sysconfig/i18n"
-
-          updated = up_to_date?(locale_file_path, new_resource.lang, new_resource.lc_all)
+          locale_file = "/etc/sysconfig/i18n"
+        elsif ::File.exist?("/etc/environment")
+          locale_file = "/etc/environment"
+        else
+          raise "#{node["platform"]} platform not supported by the chef locale resource."
+        end
 
-          file locale_file_path do
-            content lazy {
-              locale = IO.read(locale_file_path)
-              variables = Hash[locale.lines.map { |line| line.strip.split("=") }]
-              variables["LANG"] = new_resource.lang
-              variables["LC_ALL"] =
-                variables.map { |pairs| pairs.join("=") }.join("\n") + "\n"
-            }
-            not_if { updated }
-          end
+        contents = IO.readlines(locale_file)
+        env_val = contents.map { |t| t.split("=") if t.include?("=") }.compact.to_h
 
-          execute "reload root's lang profile script" do
-            command "source /etc/sysconfig/i18n; source /etc/profile.d/lang.sh"
-            not_if { updated }
-          end
-        elsif ::File.exist?("/usr/sbin/update-locale")
+        unless up_to_date?(env_val)
           execute "Generate locales" do
-            command "locale-gen"
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang, new_resource.lc_all) }
+            command "locale-gen #{new_resource.lang}"
+            only_if { locale_file == "/etc/default/locale" }
           end
 
-          execute "Update locale" do
-            command "update-locale LANG=#{new_resource.lang} LC_ALL=#{new_resource.lc_all}"
-            not_if { up_to_date?("/etc/default/locale", new_resource.lang, new_resource.lc_all) }
+          file locale_file do
+            content replace(env_val)
           end
-        else
-          raise "#{node["platform"]} platform not supported by the chef locale resource."
         end
       end
 
       action_class do
-        def up_to_date?(file_path, lang, lc_all = nil)
-          locale = IO.read(file_path)
-          locale.include?("LANG=#{lang}") &&
-            (node["init_package"] == "systemd" || lc_all.nil? || locale.include?("LC_ALL=#{lc_all}"))
-        rescue
-          false
+        def up_to_date?(hash)
+          hash["LANG"] && new_resource.lang == hash["LANG"].gsub(/\n|"/, "") &&
+            hash["LC_ALL"] && new_resource.lc_all == hash["LC_ALL"].gsub(/\n|"/, "")
+        end
+
+        def replace(hash)
+          hash["LANG"] = "\"#{new_resource.lang}\""
+          hash["LC_ALL"] = "\"#{new_resource.lc_all}\""
```

Comment:
If `LC_ALL` should only be set if the user sets a value on the resource, default to nil could guard whether LC_ALL gets written.

```ruby
hash["LC_ALL"] = "\"#{new_resource.lc_all}\"" if new_resource.lc_all
```

Also, tip, to avoid needing to escape double-quotes when interpolating in strings, you can use `%Q{}` (or only `%{}`):

```ruby
%Q{"#{new_resource.lang}"}
```

Reply:
I've updated the code. I think I've covered almost all the cases.
Not enclosing the values in double quotes. Not a breaking change so far.